fiat-roles/src/test/groovy/com/netflix/spinnaker/fiat/permissions/RedisPermissionsRepositorySpec.groovy

```diff
@@ -29,53 +29,151 @@ import com.netflix.spinnaker.fiat.model.resources.Role
 import redis.clients.jedis.Jedis
 import redis.clients.jedis.JedisPool
 import spock.lang.AutoCleanup
 import spock.lang.Shared
 import spock.lang.Specification
 import spock.lang.Subject
 
+import java.time.Clock
+import java.time.Duration
+import java.time.Instant
+import java.time.ZoneId
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.ExecutionException
+import java.util.concurrent.Executors
+
 class RedisPermissionsRepositorySpec extends Specification {
 
   private static final String EMPTY_PERM_JSON = "{}"
 
   private static final String UNRESTRICTED = UnrestrictedResourceConfig.UNRESTRICTED_USERNAME
 
-  static String prefix = "unittests"
-
   @Shared
   @AutoCleanup("destroy")
   EmbeddedRedis embeddedRedis
 
   @Shared
   ObjectMapper objectMapper = new ObjectMapper().setSerializationInclusion(JsonInclude.Include.NON_NULL)
 
+  RedisPermissionRepositoryConfigProps configProps = new RedisPermissionRepositoryConfigProps(prefix: "unittests")
+
   @Shared
   Jedis jedis
 
+  @Shared
+  PausableRedisClientDelegate redisClientDelegate
+
   @Subject
   RedisPermissionsRepository repo
 
+  Clock clock = new TestClock()
+
   def setupSpec() {
     embeddedRedis = EmbeddedRedis.embed()
     jedis = embeddedRedis.jedis
     jedis.flushDB()
+    redisClientDelegate = new PausableRedisClientDelegate(new JedisClientDelegate(embeddedRedis.pool as JedisPool))
+  }
+
+  private static class TestClock extends Clock {
+    private Instant instant = Instant.now()
+    @Override
+    ZoneId getZone() {
+      return ZoneId
+    }
+
+    @Override
+    Clock withZone(ZoneId zone) {
+      throw new UnsupportedOperationException()
+    }
+
+    @Override
+    Instant instant() {
+      return instant
+    }
+
+    void tick(Duration amount) {
+      this.instant = instant.plus(amount)
+    }
+  }
+
+  static class PausableRedisClientDelegate implements RedisClientDelegate {
+    private final RedisClientDelegate delegate
+    private CountDownLatch pauseLatch = null
+
+    PausableRedisClientDelegate(RedisClientDelegate delegate) {
+      this.delegate = delegate
+    }
+
+    void pause() {
+      pauseLatch = new CountDownLatch(1)
+    }
+
+    void resume() {
+      if (pauseLatch != null) {
+        pauseLatch.countDown()
+        pauseLatch = null
+      }
+    }
+
+    @Delegate
+    private RedisClientDelegate getDelegate() {
+      if (pauseLatch != null) {
+        pauseLatch.await()
+      }
+      return delegate
+    }
   }
 
   def setup() {
     repo = new RedisPermissionsRepository(
+        clock,
         objectMapper,
-        new JedisClientDelegate(embeddedRedis.pool as JedisPool),
+        redisClientDelegate,
         [new Application(), new Account(), new ServiceAccount(), new Role(), new BuildService()],
-        prefix,
+        configProps,
+        RetryRegistry.ofDefaults()
     )
   }
 
   def cleanup() {
     jedis.flushDB()
   }
 
+  def "should fail if timeout is exceeded"() {
```

Comment:
woof. sorry about this test I don't really like it!

Reply:
It's looks great.